### PR TITLE
Task4

### PR DESCRIPTION
**Pull request by andrei-pastukhou**

### CODE REVIEW

File: task4/mvc/src/model/storageModel.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  /**
+   * Delete item from storage. and call actionUpdate in EventObserver.
+   *
+   * @this {storage}
+   * @param {String} key.
+   */
+  removeItem(key) {
+    localStorage.removeItem(key);
```
Comment (vladislavkovaliov) [inline task4/mvc/src/model/storageModel.js]: Add logic if element doesn't exist in storage, please

File: task4/mvc/src/model/urlRequestModel.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+   * Return data from remote server.
+   *
+   * @this {urlRequest}
+   * @return {Array}
+   */
+  getResult() {
+    const sources = [];
+    // Loads all the notes.
+    for (let key in storage.getArray()) {
+      sources.push(key);
```
Comment (vladislavkovaliov) [inline task4/mvc/src/model/urlRequestModel.js]: image

I suggest there is happening something wrong.
Adding method name into sources array isn't purpose of this logic, am i right?

Try this:
```
const sources = Object.keys(storage.getArray());
```

FYI: 
`https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Object/keys`

File: task4/mvc/src/model/urlRequestModel.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+      const url = `${config.url}?sources=${source}&sortBy=${config.sortBy}&apiKey=${config.apiKey}`;
+      this.actionPending.broadcast();
+      fetch(url)
+        .then((response) => {
+          return response.json();
+        })
+        .then((data) => {
+          return data.articles;
+        }).then((articles) => {
+        this.actionFulfilled.broadcast(articles);
```
Comment (vladislavkovaliov) [inline task4/mvc/src/model/urlRequestModel.js]: Add spaces, please

File: task4/mvc/src/view/linkView.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+   *
+   * @constructor
+   * @this {linkView}
+   * @param {object} model. This view's logic for this object.
+   */
+  constructor(model) {
+    super();
+    this.model = model;
+    this.actionClick = new EventObserver();
+    this.clases = ['flex-sm-fill', 'text-sm-center', 'nav-link'];
```
Comment (vladislavkovaliov) [inline task4/mvc/src/view/linkView.js]: Replace this into `init` method, please

**Comment by vladislavkovaliov** [comment]
image

Make sure that behavior is correct, please

File: task4/mvc/src/controller/linkListController.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    this.view = view;
+  }
+
+  /**
+   * function of initialization controller. Use to implement action in eventObserver.
+   *
+   * @this {linkListController}
+   */
+  init() {
+    this.model.list.forEach((link) => {
```
[the remark this answers is not in the document]
Comment (andrei-pastukhou) [reply task4/mvc/src/controller/linkListController.js]: done

File: task4/mvc/src/controller/linkListController.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+      let linkView = new this.view(link);
+      linkView.actionClick.subscribe(() => {
```
[the remark this answers is not in the document]
Comment (andrei-pastukhou) [reply task4/mvc/src/controller/linkListController.js]: done

File: task4/mvc/src/controller/newsController.js [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+   *
+   * @this {newsController}
+   */
+  init() {
+    // Logic for actionUpdate of storage.
+    this.storage.actionUpdate.subscribe(() => {
+      this.request.getResult(this.storage);
+    });
+    // Logic for actionPending of request.
+    this.request.actionPending.subscribe(() => {
```
[the remark this answers is not in the document]
Comment (andrei-pastukhou) [reply task4/mvc/src/controller/newsController.js]: done

File: task4/mvc/src/controller/newsController.js [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+      console.log('start downloading');
+    });
+    // Logic for actionFulfilled of request.
+    this.request.actionFulfilled.subscribe((data) => {
```
[the remark this answers is not in the document]
Comment (andrei-pastukhou) [reply task4/mvc/src/controller/newsController.js]: done

File: task4/mvc/src/controller/newsController.js [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+   * function of display (redraw) if newsList.
+   *
+   * @this {newsController}
+   */
+  drawNews() {
+    // Do clear for newsList dom.
+    this.newsList.domElement.innerHTML = '';
+
+    // For each newsModel in list create a view and append it to list dom.
+    if(this.newsList.list) {
```
[the remark this answers is not in the document]
Comment (andrei-pastukhou) [reply task4/mvc/src/controller/newsController.js]: done

File: task4/mvc/src/model/listModel.js [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+   * @param {class} childClass. class of object's model. example: linkModel, newsModel.
+   * @param {array} modelsArray. array with data to model of childClass. According with type of  childClass structure of
+   *  object inside array can be different.
+   * @param {object} options. Object with parameter's to list model. Give it from config. example {id: id_element_of_dom}
+   *  use to define dom element where will put a listing of this list.
+   */
+  constructor(childClass, modelsArray, options) {
+    this.childClass = childClass;
+    this.list = [];
+    if(modelsArray) {
```
[the remark this answers is not in the document]
Comment (andrei-pastukhou) [reply task4/mvc/src/model/listModel.js]: done

File: task4/mvc/src/model/listModel.js [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+  /**
+   * Update (before delete than paste) object's list (this.list) with new object of this.childClass class according with
+   * array data. (array data is containers of parameters fore each object)
+   *
+   * @this {listModel}
+   * @param {array} data
+   */
+  updateData(data) {
+    this.list = [];
+    if(data) {
```
[the remark this answers is not in the document]
Comment (andrei-pastukhou) [reply task4/mvc/src/model/listModel.js]: done

File: task4/mvc/src/model/storageModel.js (hunk 1 above)
Comment (andrei-pastukhou) [reply task4/mvc/src/model/storageModel.js]: done

File: task4/mvc/src/model/urlRequestModel.js (hunk 2 above)
Comment (andrei-pastukhou) [reply task4/mvc/src/model/urlRequestModel.js]: done: Changed getArray function. to : 
  static getArray(keyWord = false) {
    const sources = Object.keys(localStorage);
    if(!keyWord) {
      return sources;
    }
    let returnArray = [];
    if(Array.isArray(sources) && (sources.length > 0)){
      sources.forEach((item) => {
        if(localStorage.getItem(item) === keyWord) {
          returnArray.push(item);
        }
      });
    }
    return returnArray;
  }
}

File: task4/mvc/src/model/urlRequestModel.js (hunk 3 above)
Comment (andrei-pastukhou) [reply task4/mvc/src/model/urlRequestModel.js]: done

File: task4/mvc/src/view/linkView.js (hunk 4 above)
Comment (andrei-pastukhou) [reply task4/mvc/src/view/linkView.js]: done

File: task4/redux/src/actions/index.js [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+    let sources = [];
+    currentState.channels.map((channel) => {
+      if (channel.selected){
+        sources.push(channel.source);
+      }});
+    const source = sources.join(',');
+    if(source){
+      dispatch({type: FETCH_NEWS});
+      const req = new Request(`${config.url}?sources=${source}&sortBy=publishedAt&apiKey=${config.apiKey}`);
+      fetch(req)
```
Comment (vladislavkovaliov) [inline task4/redux/src/actions/index.js]: Try just 
```
fetch(`${config.url}?sources=${source}&sortBy=publishedAt&apiKey=${config.apiKey}`)
```
without `Request`, please

File: task4/redux/src/actions/index.js [hunk 13]
```diff
@@ -0,0 +1,40 @@
+import {config} from '../config.js'
+
+//define constants for action.type.
+export const FETCH_NEWS = 'FETCH_NEWS';
```
Comment (vladislavkovaliov) [inline task4/redux/src/actions/index.js]: Export these actions into single file, please

Comment (andrei-pastukhou) [reply task4/redux/src/actions/index.js]: done. Define of action type's constants moved to ./contstants/index.js file.

File: task4/redux/src/actions/index.js (hunk 12 above)
Comment (andrei-pastukhou) [reply task4/redux/src/actions/index.js]: done.

**Pull request merged**